### PR TITLE
Align score test cases

### PR DESCRIPTION
**Pull request by sidravi1**
Reviewer: @Tanmay-97 

Estimate: 1 hrs

----

# Added new test cases for guardrails to ensure llm response is consistent with context provided

## Ticket

Fixes: [AAQ-252](https://idinsight.atlassian.net/browse/AAQ-252)

## Description, Motivation and Context

In the spirit of Test-Driven-Development, I wanted to raise this PR with test cases for the guardrail to check.

@Tanmay-97 - Would you be able to look through these and see if you agree with the "True" and "False" labels I have provided? 

### Changes

Test data and new placeholder test cases.

## How Has This Been Tested?

You can run the `read_test_data` function but really no testing is needed since it is mostly just data.

## Next up
1. Guardrails functionality using GPT-4
2. Guardrails functionality using AlignScore
3. Docs

## Checklist:

This checklist is a useful reminder of small things that can easily be forgotten.
Put an `x` in all the items that apply and remove any items that are not relevant to this PR.

- [x] My code follows the style guidelines of this project
- [x] I have reviewed my own code to ensure good quality
- [x] I have tested the functionality of my code to ensure it works as intended
- [x] I have resolved merge conflicts
- [x] I have updated the automated tests (if applicable)
- [x] I have written [good commit messages][1]
- [ ] I have updated the README file (if applicable)
- [ ] I have updated affected documentation (if applicable)

### CODE REVIEW

**Review by Tanmay-97 — APPROVED** [review]
Thanks @sidravi1 , left a few comments.

At a high-level, how would these rails help make LLM responses safer? 
Because the test cases are like GRE's Reading Comprehension questions. We know already that GPT-4 passes that with flying colors.
I need to read a bit more on guardrails, but we could we think about corner/edge cases _in the context__ where the LLM operates e.g. maternal health

Also, I wonder if it's worth introducing trinary logic. Some question are better answered with `unclear from the para` instead of `True` or `False`

File: core_backend/tests/rails/data/llm_response_in_content.yaml [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+    living with diabetes, access to affordable treatment, including insulin, is
+    critical to their survival. There is a globally agreed target to halt the rise in
+    diabetes and obesity by 2025.\nAbout 422 million people worldwide have diabetes,
+    the majority living in low-and middle-income countries, and 1.5 million deaths are
+    directly attributed to diabetes each year. Both the number of cases and the
+    prevalence of diabetes have been steadily increasing over the past few decades
+  tests:
+    - statement: being overweight causes diabetes
+      expected: False
+      reason: |
```
[the remark this answers is not in the document]
Comment (sidravi1) [reply core_backend/tests/rails/data/llm_response_in_content.yaml:19]: The `reason` field is just to display when the test fails so we know why we chose the label we did.

File: core_backend/tests/rails/data/llm_response_in_content.yaml [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+        This statement is too strong. Being overweight CAN cause diabetes but does this
+        is not definite
+    - statement: being overweight can cause diabetes
+      expected: True
+      reason: |
```
[the remark this answers is not in the document]
Comment (sidravi1) [reply core_backend/tests/rails/data/llm_response_in_content.yaml:24]: Thanks! You are right.

File: core_backend/tests/rails/data/llm_response_in_content.yaml [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        It mentions 422 million with diabetes and MAJORITY living in LMICs
+    - statement: There are more than 200 million people within diabetes in low and middle income countrie
+      expected: True
+      reason: |
+        It mentions 422 million with diabetes and MAJORITY living in LMICs
+    - statement: There are more than 150 million people within diabetes in low and middle income countrie
+      expected: False
+      reason: |
+        It mentions 422 million with diabetes and MAJORITY living in LMICs.
+        150 is not a majority.
```
[the remark this answers is not in the document]
Comment (sidravi1) [reply core_backend/tests/rails/data/llm_response_in_content.yaml:46]: Ah. It should be 350 not 150 here. I'll fix it.

File: core_backend/tests/rails/data/llm_response_in_content.yaml [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+      reason: |
+        it mentions 'increasing over the past few decades'.
+    - statement: the treatment for diabetes can be expensive
+      expected: True
+      reason: |
+        It implies affordable treatment as a barrier
+    - statement: the treatment for diabetes is expensive
+      expected: False
+      reason: |
+        It implies affordable treatment as a barrier but not necessarily for everyone.
```
[the remark this answers is not in the document]
Comment (sidravi1) [reply core_backend/tests/rails/data/llm_response_in_content.yaml:72]: Thanks. I'll update this

File: core_backend/tests/rails/data/llm_response_in_content.yaml [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+    or diabetes mellitus in pregnancy, according to WHO criteria. Health-care providers
+    should ask all pregnant women about their tobacco use (past and present) and exposure
+    to second-hand smoke as early as possible in pregnancy and at every antenatal care
+    visit. Health-care providers should ask all pregnant women about their
+    use of alcohol and other substances (past and present) as early as possible in the
+    pregnancy and at every antenatal care visit.
+  tests:
+    - statement: Full blood count is not the best way for diagnosing anaemia
+      expected: True
+      reason: Full blood count testing is the recommended method for diagnosing ...
```
[the remark this answers is not in the document]
Comment (sidravi1) [reply core_backend/tests/rails/data/llm_response_in_content.yaml:227]: Good point. I'll change it to "recommended way" instead of "best way"

File: core_backend/tests/rails/data/llm_response_in_content.yaml [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+    - statement: Hameloglobin testing is the best way for diagnosing anaemia
+      expected: False
+      reason: Full blood count testing is the recommended method for diagnosing ...
```
[the remark this answers is not in the document]
Comment (sidravi1) [reply core_backend/tests/rails/data/llm_response_in_content.yaml:230]: True! I'll also update this

File: core_backend/tests/rails/data/llm_response_in_content.yaml [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+        diagnosing anaemia
+      expected: True
+      reason: |
+        haemoglobin testing with a haemoglobinometer is recommended over
+        the use of the haemoglobin colour scale
+    - statement: |
+        The order of preference in methods for diagnosing ASB is: midstream urine
+        culture, midstream urine gram-staining, dipstick tests.
+      expected: True
+      reason: midstream urine gram-staining > midstream urine culture > dipstick test
```
[the remark this answers is not in the document]
Comment (sidravi1) [reply core_backend/tests/rails/data/llm_response_in_content.yaml:245]: Thanks! Will fix

File: core_backend/tests/rails/data/llm_response_in_content.yaml [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+    - statement: |
+        The order of preference in methods for diagnosing ASB is: midstream urine
+        gram-staining, midstream urine culture, dipstick tests.
+      expected: False
+      reason: midstream urine gram-staining > midstream urine culture > dipstick test
```
[the remark this answers is not in the document]
Comment (sidravi1) [reply core_backend/tests/rails/data/llm_response_in_content.yaml:250]: Thanks! Will fix

File: core_backend/tests/rails/data/llm_response_in_content.yaml [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+      reason: should be done where there is capacity and WHO requirements are met
+    - statement: |
+        Clinical enquiry about the possibility of intimate partner violence (IPV) should
+        be done where there is capacity
+      expected: False
+      reason: should be done where there is capacity AND WHO requirements are met
+    - statement: |
+        Clinical enquiry about the possibility of intimate partner violence (IPV) should
+        be done where minimum WHO requirement are met.
+      expected: True
```
[the remark this answers is not in the document]
Comment (sidravi1) [reply core_backend/tests/rails/data/llm_response_in_content.yaml:264]: Thanks! Will update

**Comment by sidravi1** [comment]
> At a high-level, how would these rails help make LLM responses safer? Because the test cases are like GRE's Reading Comprehension questions. We know already that GPT-4 passes that with flying colors. I need to read a bit more on guardrails, but we could we think about corner/edge cases _in the context__ where the LLM operates e.g. maternal health

So what these do is make sure LLM is only answering from the context and not using it's pre-training or other sources. So for example, if the context provided is in conflict with it's pre-training, it should prefer context over pre-training. This give me an idea for another test case with made up context that is wrong and see if it adheres to it.

> Also, I wonder if it's worth introducing trinary logic. Some question are better answered with `unclear from the para` instead of `True` or `False`

If it is unclear, then it should say "False". We don't want the LLM response to say things that it is not confident about. What "confidence" means will be a parameter. Look out for the next PR!
